WIP - MTA-4637: Pattern matching docs need urgent improvements (#92)
migtools/mta-documentation: docs/topics/yaml-builin-provider.adoc

// Module included in the following assemblies:
//
// * docs/rules-development-guide/master.adoc

:_mod-docs-content-type: REFERENCE
[id="yaml-builin-provider_{context}"]
= Builtin provider

> The `builtin` is an internal provider that can analyze various files and internal metadata generated by the engine.
> 
> This provider has the following capabilities:

Reviewer: ```suggestion
The `builtin` is an internal provider that can analyze various files and internal metadata generated by the engine. This provider has the following capabilities:
```

* `file`
* `filecontent`
* `xml`
* `json`
* `hasTags`

.`file`

> The `file` capability enables the provider to search for files in the source code that match a given pattern.

Reviewer: ```suggestion
By using the `file` capability, the provider searches for files in the source code that match a given pattern.
```

[source,yaml]
----
when:
  builtin.file:
    pattern: "<regex_to_match_filenames>"
----

.`filecontent`

> The `filecontent` capability enables the provider to search for content that matches a given pattern.

Reviewer: ```suggestion
By using the `filecontent` capability, the provider searches for content that matches a given pattern.
```

[source,yaml]
----
when:
  builtin.filecontent:
    filePattern: "<regex_to_match_filenames_to_scope_search>"
    pattern: "<regex_to_match_content_in_the_matching_files>"
----

.`xml`

> The `xml` capability enables the provider to query XPath expressions on a list of provided XML files. This capability takes 2 input parameters, `xpath` and `filepaths`.

Reviewer: ```suggestion
By using the `xml` capability, the provider queries XPath expressions on a list of provided XML files. This capability takes 2 input parameters, `xpath` and `filepaths`.
```

[source,yaml]
----
when:
  builtin.xml:
    xpath: "<xpath_expressions>" # <1>
    filepaths: # <2>
      - "/src/file1.xml"
      - "/src/file2.xml"
----
<1> `xpath` must be a valid XPath expression.
<2> `filepaths` is a list of files to apply the XPath query to.

.`json`

> The `json` capability enables the provider to query XPath expressions on a list of provided JSON files. Currently, `json` only takes XPath as input and performs the search on all JSON files in the codebase.

Reviewer: ```suggestion
By using the `json` capability, the provider queries XPath expressions on a list of provided JSON files. Currently, `json` only takes XPath as input and performs the search on all JSON files in the codebase.
```

[source,yaml]
----
when:
  builtin.json:
    xpath: "<xpath_expressions>" # <1>
----
<1> `xpath` must be a valid XPath expression.

.`hasTags`

> The `hasTags` capability enables the provider to query application tags. It queries the internal data structure to check whether the application has the given tags.

Reviewer: ```suggestion
By using the `hasTags` capability, the provider queries application tags. It queries the internal data structure to check whether the application has the given tags.
```
